src/node/txorphanage.cpp

```diff
@@ -8,68 +8,201 @@
 
-    using OrphanMap = decltype(m_orphans);
-
-    struct IteratorComparator
+    // Index by wtxid, then peer
+    struct ByWtxid {};
+    using ByWtxidView = std::tuple<Wtxid, NodeId>;
+    struct WtxidExtractor
     {
-        template<typename I>
-        bool operator()(const I& a, const I& b) const
+        using result_type = ByWtxidView;
+        result_type operator()(const Announcement& ann) const
         {
-            return a->first < b->first;
+            return ByWtxidView{ann.m_tx->GetWitnessHash(), ann.m_announcer};
         }
     };
 
-    /** Index from the parents' COutPoint into the m_orphans. Used
-     *  to remove orphan transactions from the m_orphans */
-    std::map<COutPoint, std::set<OrphanMap::iterator, IteratorComparator>> m_outpoint_to_orphan_it;
+    // Sort by peer, then by whether it is ready to reconsider, then by recency.
+    struct ByPeer {};
+    using ByPeerView = std::tuple<NodeId, bool, SequenceNumber>;
+    struct ByPeerViewExtractor {
+        using result_type = ByPeerView;
+        result_type operator()(const Announcement& ann) const
+        {
+            return ByPeerView{ann.m_announcer, ann.m_reconsider, ann.m_entry_sequence};
+        }
+    };
+
+    struct OrphanIndices final : boost::multi_index::indexed_by<
+        boost::multi_index::ordered_unique<boost::multi_index::tag<ByWtxid>, WtxidExtractor>,
+        boost::multi_index::ordered_unique<boost::multi_index::tag<ByPeer>, ByPeerViewExtractor>
+    >{};
+
+    using AnnouncementMap = boost::multi_index::multi_index_container<Announcement, OrphanIndices>;
+    template<typename Tag>
+    using Iter = typename AnnouncementMap::index<Tag>::type::iterator;
+    AnnouncementMap m_orphans;
+
+    const TxOrphanage::Count m_max_global_latency_score{DEFAULT_MAX_ORPHANAGE_LATENCY_SCORE};
+    const TxOrphanage::Usage m_reserved_usage_per_peer{DEFAULT_RESERVED_ORPHAN_WEIGHT_PER_PEER};
+
+    /** Number of unique orphans by wtxid. Less than or equal to the number of entries in m_orphans. */
+    TxOrphanage::Count m_unique_orphans{0};
+
+    /** Memory used by orphans (see Announcement::GetMemUsage()), deduplicated by wtxid. */
+    TxOrphanage::Usage m_unique_orphan_bytes{0};
+
+    /** The sum of each unique transaction's latency scores including the inputs only (see Announcement::GetLatencyScore
+     * but subtract 1 for the announcements themselves). The total orphanage's latency score is given by this value +
+     * the number of entries in m_orphans. */
+    TxOrphanage::Count m_unique_rounded_input_scores{0};
 
-    /** Orphan transactions in vector for quick random eviction */
-    std::vector<OrphanMap::iterator> m_orphan_list;
+    /** Index from the parents' outputs to wtxids that exist in m_orphans. Used to find children of
+     * a transaction that can be reconsidered and to remove entries that conflict with a block.*/
+    std::unordered_map<COutPoint, std::set<Wtxid>, SaltedOutpointHasher> m_outpoint_to_orphan_it;
```

Comment:
in d5097d6eef85eab971352cf78fb94a30c6e6f127: nit: the naming seems outdated, considering that (in contrast to master) the map's value type is not an iterator anymore?
```suggestion
    std::unordered_map<COutPoint, std::set<Wtxid>, SaltedOutpointHasher> m_outpoint_to_orphan_wtxids;
```

Reply:
Added to followup